File: src/Hemordna.Client/wwwroot/js/scroll-state.js
// Toggles document.documentElement.dataset.scrolled so CSS elsewhere (NavMenu.razor.css's
// compact pill, Pages/MinDag.razor.css's/Hushall.razor.css's collapsed header row) can react to
// scroll position without its own per-page listener. CSS's animation-timeline: scroll() would
// remove the need for this entirely, but lacks Firefox support - a data attribute plus plain CSS
// transitions works everywhere.
const threshold = 24;

let rafId = null;

function onScroll() {
    if (rafId !== null) {
        return;
    }

    rafId = requestAnimationFrame(() => {
        rafId = null;

        if (window.scrollY > threshold) {
            document.documentElement.dataset.scrolled = '';
        } else {
            delete document.documentElement.dataset.scrolled;
        }
    });
}

export function attach() {
    window.addEventListener('scroll', onScroll, { passive: true });
}

// A Blazor route change never reloads the page, so the browser never runs its own "new page"
// scroll reset - a scrollY left over from whatever page the member was on before (even a few px,
// enough to cross the threshold above without looking "scrolled" at all) would otherwise survive
// straight into the newly rendered page. Called from MainLayout on every navigation
// (Support/ScrollState.cs) - moves the actual scroll position, not just the attribute, so the
// underlying state is genuinely correct rather than a flag masking a stale scrollY.
export function reset() {
    window.scrollTo(0, 0);
    delete document.documentElement.dataset.scrolled;
}

export function dispose() {
    window.removeEventListener('scroll', onScroll);

    if (rafId !== null) {
        cancelAnimationFrame(rafId);
        rafId = null;
    }

    delete document.documentElement.dataset.scrolled;
}

File: src/Hemordna.Client/Support/ScrollState.cs
using Microsoft.JSInterop;

namespace Hemordna.Client.Support;

/// <summary>
/// Toggles <c>data-scrolled</c> on <c>&lt;html&gt;</c> as the page scrolls (app.css/NavMenu.razor.css/
/// Pages/MinDag.razor.css read it from there) - a single, app-wide listener attached once from
/// <c>MainLayout.OnAfterRenderAsync(firstRender)</c> rather than one per page.
/// </summary>
internal static class ScrollState
{
    private static IJSObjectReference? _module;

    public static async Task AttachAsync(IJSRuntime js)
    {
        _module = await js.InvokeAsync<IJSObjectReference>("import", "./js/scroll-state.js");
        await _module.InvokeVoidAsync("attach");
    }

    /// <summary>
    /// Clears a stale <c>data-scrolled</c> left over from wherever the member was before a
    /// route change - see <c>MainLayout.razor</c>'s own <c>LocationChanged</c> subscription and
    /// docs/ARCHITECTURE.md "Uppdrag: fokuskortet". A no-op before <see cref="AttachAsync"/> has
    /// run (an early redirect, e.g. <c>Redirect.razor</c>, can navigate before
    /// <c>MainLayout.OnAfterRenderAsync(firstRender)</c> gets a chance to attach the module).
    /// </summary>
    public static async Task ResetAsync()
    {
        if (_module is not null)
        {
            await _module.InvokeVoidAsync("reset");
        }
    }
}
